### ISSUE TITLE
terraform fmt on save not actually occurring

### ISSUE THREAD

**Issue by asht0nium** [hashicorp/vscode-terraform#167]
If have Format on Save enabled, and on saving a .tf file, the output shows it running `terraform fmt` on the file:

```
2019-03-29 15:15:41:274 [INFO ] formatting-provider: running 'terraform fmt' on 'c:\Users\user\project-dir\main.tf'
2019-03-29 15:15:41:275 [INFO ] runner: Running terraform cwd='C:\Users\user\AppData\Local\Programs\Microsoft VS Code' path='terraform' (version: 0.11.11) args=[fmt, -]
2019-03-29 15:15:42:492 [INFO ] file-system-watcher: Indexed file:///c%3A/Users/user/project-dir/main.tf
2019-03-29 15:15:44:340 [INFO ] runner: Running terraform succeeded.
```

However, the formatting is not applied and file remains unchanged.

Running `terraform fmt` on the file via the console has the expected formatting effect on the file.

_This issue was also mentioned by @IronFarm in https://github.com/mauve/vscode-terraform/issues/62#issuecomment-426944230_

**Comment by askldjd** [comment]
I am seeing something very similar as well.

```
2019-05-30 01:25:10:462 [INFO ] formatting-provider: running 'terraform fmt' on '/home/askldjd/work/code/test.tf'
2019-05-30 01:25:10:462 [INFO ] runner: Running terraform cwd='/home/askldjd' path='/bin/terraform' (version: 0.12.0) args=[fmt, -]
2019-05-30 01:25:10:515 [ERROR] runner: Running terraform failed: Error: Command failed: /bin/terraform fmt -
2019-05-30 01:25:10:515 [ERROR] runner: [31m
2019-05-30 01:25:10:515 [ERROR] runner: [1m[31mError: [0m[0m[1mInvalid argument name[0m
2019-05-30 01:25:10:515 [ERROR] runner: 
2019-05-30 01:25:10:515 [ERROR] runner: [0m  on <stdin> line 56:
2019-05-30 01:25:10:515 [ERROR] runner:   (source code not available)
2019-05-30 01:25:10:515 [ERROR] runner: 
2019-05-30 01:25:10:515 [ERROR] runner: Argument names must not be quoted.
2019-05-30 01:25:10:515 [ERROR] runner: [0m[0m
2019-05-30 01:25:10:515 [ERROR] runner: 
2019-05-30 01:25:10:515 [ERROR] formatting-provider: formatting failed. 
Error: Invalid argument name
```

I am on
```
Version: 1.34.0
Commit: a622c65b2c713c890fcf4fbf07cf34049d5fe758
Date: 2019-05-21T12:30:20.933Z
Electron: 4.1.4
Chrome: 69.0.3497.128
Node.js: 10.11.0
V8: 6.9.427.30-electron.0
OS: Linux x64 4.19.45-1-lts
```

with Terraform extension version 1.3.11

**Comment by mauve** [comment]
The log you posted says that you have syntax errors in the file (it might be picking the wrong terraform version for running `terraform fmt`, are your templates really compatible with the version of terraform at path `/bin/terraform`?). When there are syntax errors `terraform fmt -` fails with an error code and the code is not reformated.

**Comment by askldjd** [comment]
Here is the thing I don't understand. There was no error from `terraform fmt`.

**Comment by udondan** [comment]
Anyone found a solution or knows the cause?  I'm having this problem since some weeks and can't find a solution. The Terraform output window shows it is running but nothing happens with the document.

```
2019-11-18 09:01:37:238 [INFO ] formatting-provider: running 'terraform fmt' on '/path/to/file.tf'
2019-11-18 09:01:37:238 [INFO ] runner: Running terraform cwd='/' path='/usr/local/bin/terraform' (version: 0.12.15) args=[fmt, -]
2019-11-18 09:01:38:189 [INFO ] runner: Running terraform succeeded.
```

I'm confused about the args: `args=[fmt, -]`

Is the 2nd arg a placeholder for the filename? Or is that the issue?

**Comment by IronFarm** [comment]
"-" usually means "read the input from STDIN". I'd expect that the extension is reading the file in, running the terraform format command `terraform fmt -` and piping the file contents into the process.

This seems a reasonable way of achieving file formatting. Are you sure it isn't working for you? The challenge here is figuring out _what_ is getting piped into the formatting command.

**Comment by udondan** [comment]
> "-" usually means "read the input from STDIN". I'd expect that the extension is reading the file in, running the terraform format command terraform fmt - and piping the file contents into the process.

Yes, that's correct, that's what it's doing. 

Debuggin it right now. The input is correct, the output even is the correctly formatted code. The problem is, [this line](https://github.com/mauve/vscode-terraform/blob/master/src/format.ts#L35) doesn't change the document, even though the result of that line looks good to me.

```json
{
    "range": [
        {
            "line": 0,
            "character": 0
        },
        {
            "line": 10,
            "character": 0
        }
    ],
    "newText": "resource \"aws_organizations_account\" \"this\" {\n  count    = local.env == \"master\" ? 0 : 1\n  provider = aws.master\n  name     = local.env\n  email    = local.config.account.email\n}\n\noutput \"account\" {\n  value = local.account\n}\n"
}
```

The `newText` is correctly formatted/indended and a newline is added to the end of the file, which I messed up for this test case.

**Comment by udondan** [comment]
Can't figure it out... I just replaced the `-` now with `document.fileName`. Certainly not the way to do it and sometimes causes issues with vsc complaining the file it wants to save is newer than what it got. Better though than not formatting the code.

BTW, manually formatting the code works. It's only the format on save that doesn't do anything, event though the function is triggered and debugging output looks as if it worked.

PS: How do i have version 1.4.0 installed when the latest release here is 1.3.12?

**Comment by paultyng** [comment]
We just released [v2.0.0-rc.1](https://github.com/hashicorp/vscode-terraform/releases/tag/v2.0.0-rc.1) of the extension. The main features include:

* Added syntax support for 0.12
* Added [terraform-ls](https://github.com/hashicorp/terraform-ls) usage by default (currently on 0.3.0, which offers basic provider code completion)

You can find additional information and specifics in the [release notes](https://github.com/hashicorp/vscode-terraform/releases/tag/v2.0.0-rc.1) and [CHANGELOG](https://github.com/hashicorp/vscode-terraform/blob/master/CHANGELOG.md).

With this release we expect that many of the prior issues and PRs are no longer relevant or have been addressed, and are therefore being closed. If you feel the action taken on an issue or PR is in error, please comment as such and we can figure out the appropriate way to address it.

We plan to add the final 2.0.0 release to the marketplace soon, but are actively seeking your feedback now on the release candidates. You can download the .vsix from the releases page and [manually install it in VS Code](https://code.visualstudio.com/docs/editor/extension-gallery#_install-from-a-vsix) to try it out.
